test/framework/control_plane.go

```diff
@@ -34,213 +34,276 @@ import (
-	Expect(err).NotTo(HaveOccurred(), "stack: %+v", err)
+		// TODO: Return status.Ready instead...
+		return controlplane.Status.Initialized, nil
+	}, intervals...).Should(BeTrue())
+}
+
+// DeleteClusterInput is the input for DeleteCluster.
+type DeleteClusterInput struct {
+	Deleter Deleter
+	Cluster *clusterv1.Cluster
+}
 
+// DeleteCluster deletes the cluster and waits for everything the cluster owned to actually be gone.
+func DeleteCluster(ctx context.Context, input DeleteClusterInput) {
 	By(fmt.Sprintf("deleting cluster %s", input.Cluster.GetName()))
-	Expect(mgmtClient.Delete(ctx, input.Cluster)).To(Succeed())
+	Expect(input.Deleter.Delete(ctx, input.Cluster)).To(Succeed())
+}
 
+// WaitForClusterDeletedInput is the input for WaitForClusterDeleted.
+type WaitForClusterDeletedInput struct {
+	Getter  Getter
+	Cluster *clusterv1.Cluster
+}
+
+// WaitForClusterDeleted waits until the cluster object has been deleted.
+func WaitForClusterDeleted(ctx context.Context, input WaitForClusterDeletedInput, intervals ...interface{}) {
+	By(fmt.Sprintf("waiting for cluster %s to be deleted", input.Cluster.GetName()))
 	Eventually(func() bool {
-		clusters := clusterv1.ClusterList{}
-		if err := mgmtClient.List(ctx, &clusters); err != nil {
-			fmt.Println(err.Error())
-			return false
+		cluster := &clusterv1.Cluster{}
+		key := client.ObjectKey{
+			Namespace: input.Cluster.GetNamespace(),
+			Name:      input.Cluster.GetName(),
 		}
-		return len(clusters.Items) == 0
-	}, input.DeleteTimeout, eventuallyInterval).Should(BeTrue())
+		return apierrors.IsNotFound(input.Getter.Get(ctx, key, cluster))
+	}, intervals...).Should(BeTrue())
+}
+
+// AssertAllClusterAPIResourcesAreGoneInput is the input for AssertAllClusterAPIResourcesAreGone.
+type AssertAllClusterAPIResourcesAreGoneInput struct {
+	Lister  Lister
+	Cluster *clusterv1.Cluster
+}
 
+// AssertAllClusterAPIResourcesAreGone ensures that all known Cluster API resources have been remvoed.
+func AssertAllClusterAPIResourcesAreGone(ctx context.Context, input AssertAllClusterAPIResourcesAreGoneInput) {
 	lbl, err := labels.Parse(fmt.Sprintf("%s=%s", clusterv1.ClusterLabelName, input.Cluster.GetClusterName()))
 	Expect(err).ToNot(HaveOccurred())
-	listOpts := &client.ListOptions{LabelSelector: lbl}
+	opt := &client.ListOptions{LabelSelector: lbl}
 
 	By("ensuring all CAPI artifacts have been deleted")
-	ensureArtifactsDeleted(ctx, mgmtClient, listOpts)
-}
 
-func ensureArtifactsDeleted(ctx context.Context, mgmtClient client.Client, opt client.ListOption) {
-	// assertions
 	ml := &clusterv1.MachineList{}
-	Expect(mgmtClient.List(ctx, ml, opt)).To(Succeed())
+	Expect(input.Lister.List(ctx, ml, opt)).To(Succeed())
 	Expect(ml.Items).To(HaveLen(0))
 
 	msl := &clusterv1.MachineSetList{}
-	Expect(mgmtClient.List(ctx, msl, opt)).To(Succeed())
+	Expect(input.Lister.List(ctx, msl, opt)).To(Succeed())
 	Expect(msl.Items).To(HaveLen(0))
 
 	mdl := &clusterv1.MachineDeploymentList{}
```

Comment:
Should we also check for `clusterv1.MachinePoolList` here as well?

Reply:
Yep, will file an issue to add MachinePools to the delete list https://github.com/kubernetes-sigs/cluster-api/issues/2310